src/FormCollector.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Form;
+
+use Exception;
+use InvalidArgumentException;
+use ReflectionAttribute;
+use ReflectionNamedType;
+use ReflectionObject;
+use Yiisoft\Validator\RuleInterface;
+
+final class FormCollector
+{
+    private array $attributes;
+    private array $rules = [];
+
+    public function __construct(private FormModelInterface $formModel)
+    {
+        [$this->attributes, $this->rules] = $this->collectorAttributes();
```

Comment:
Isn't it better to do it in each any public methods to prevent collecting rules and attributes at the creation time? Kind of lazy initialization.